File: packages/nuget/src/PatternFilter.cs

using System;
using System.Collections.Generic;
using System.Linq;

namespace DesignPatterns.Catalog;

public static class PatternFilter
{
    public static IReadOnlyList<Pattern> FilterPatterns(IEnumerable<Pattern> source, string query, string category)
    {
        return source.Where(pattern =>
                (string.Equals(category, "All", StringComparison.OrdinalIgnoreCase) || string.Equals(pattern.Category, category, StringComparison.OrdinalIgnoreCase)) &&
                MatchesQuery(pattern, query))
            .ToList();
    }

    public static bool MatchesQuery(Pattern pattern, string query)
    {
        var needle = (query ?? string.Empty).Trim();
        if (needle.Length == 0)
        {
            return true;
        }

        return pattern.Name.IndexOf(needle, StringComparison.OrdinalIgnoreCase) >= 0 ||
               pattern.Intent.IndexOf(needle, StringComparison.OrdinalIgnoreCase) >= 0 ||
               pattern.UseWhen.IndexOf(needle, StringComparison.OrdinalIgnoreCase) >= 0 ||
               pattern.Category.IndexOf(needle, StringComparison.OrdinalIgnoreCase) >= 0;
    }

    public static IReadOnlyDictionary<string, int> CountByCategory(IEnumerable<Pattern> source)
    {
        var counts = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
        {
            [PatternCategory.Creational] = 0,
            [PatternCategory.Structural] = 0,
            [PatternCategory.Behavioral] = 0,
        };

        foreach (var pattern in source)
        {
            counts[pattern.Category]++;
        }

        return counts;
    }
}

File: packages/java/src/main/java/com/designpatterns/PatternFilter.java

package com.designpatterns;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class PatternFilter {
    private PatternFilter() {
    }

    public static List<Pattern> filterPatterns(List<Pattern> patterns, String query, String category) {
        List<Pattern> filtered = new ArrayList<>();
        for (Pattern pattern : patterns) {
            if (("All".equalsIgnoreCase(category) || pattern.getCategory().equalsIgnoreCase(category))
                    && matchesQuery(pattern, query)) {
                filtered.add(pattern);
            }
        }
        return filtered;
    }

    public static boolean matchesQuery(Pattern pattern, String query) {
        String needle = query == null ? "" : query.trim().toLowerCase(Locale.ROOT);
        if (needle.isEmpty()) {
            return true;
        }

        return pattern.getName().toLowerCase(Locale.ROOT).contains(needle)
                || pattern.getIntent().toLowerCase(Locale.ROOT).contains(needle)
                || pattern.getUseWhen().toLowerCase(Locale.ROOT).contains(needle)
                || pattern.getCategory().toLowerCase(Locale.ROOT).contains(needle);
    }

    public static Map<String, Integer> countByCategory(List<Pattern> patterns) {
        Map<String, Integer> counts = new LinkedHashMap<>();
        counts.put(PatternCategory.CREATIONAL.value(), 0);
        counts.put(PatternCategory.STRUCTURAL.value(), 0);
        counts.put(PatternCategory.BEHAVIORAL.value(), 0);

        for (Pattern pattern : patterns) {
            counts.computeIfPresent(pattern.getCategory(), (key, value) -> value + 1);
        }

        return counts;
    }
}
